### PR TITLE
Improved cache API

### PR DESCRIPTION
**Pull request by mariomac**
#### Description of the changes

- Deprecated global cache
- Divided Cache in interface/implementation
- Allowed configuring cache in IntegrationBuilder

### Author

- [ ] add a risk label after carefully considering the "blast radius" of your changes
- [ ] describe the _intent_ of your changes in the description. don't just rewrite your code in prose
- [ ] assign at least one reviewer

### Reviewer

- [ ] review code for readability
- [ ] verify that high risk behavior changes are well tested
- [ ] check license for any new external dependency
- [ ] ask questions about anything that isn't clear and obvious
- [ ] approve the PR when you consider it's good to merge

### CODE REVIEW

File: cache/cache.go [hunk 1]
```diff
@@ -21,30 +23,33 @@ func SetNow(newNow func() time.Time) {
 
 const cacheTTL = 1 * time.Minute
 
-// Cache is a map-like structure that is initialized and stored in a JSON
-// file. It also saves the timestamp when a key was stored.
-type Cache struct {
+// Cache is a key-value structure that is initialized and stored in a persistent device.
+// It also saves the timestamp when a key was stored.
+type Cache interface {
```
Comment (varas) [inline cache/cache.go]: `Cacher` would be a better Golang idiomatic name

File: cache/cache.go [hunk 2]
```diff
@@ -78,9 +83,8 @@ func NewCache(l log.Logger) (*Cache, error) {
 	return cache, nil
 }
 
-// Save marshalls and stores the data a Cache is holding into disk as a JSON
-// file.
-func (cache *Cache) Save() error {
+// Save persists all the data in the cache.
+func (cache *cacheImpl) Save() error {
```
Comment (varas) [inline cache/cache.go]: how about `cache *cacheImpl` => `c *Cache` ?

Golang methods receivers tend to be 1or2 chars long

File: cache/cache_test.go [hunk 3]
```diff
@@ -1,84 +1,55 @@
-package cache
+package cache_test
```
Comment (varas) [inline cache/cache_test.go]: same as https://github.com/newrelic/infra-integrations-sdk/pull/59#discussion_r181799985

File: cache/global.go [hunk 4]
```diff
@@ -1,37 +1,82 @@
...
+const integrationsDir = "nr-integrations"
+const cacheDirFilePerm = 0755
+
+// DefaultPath returns a default folder/filename path to a cache for an integration from the given name. The name of
+// the file will be the name of the integration with the .json extension.
+func DefaultPath(integrationName string) string {
+	baseDir := filepath.Join(os.TempDir(), integrationsDir)
+	// Create integrations cache directory
+	err := os.MkdirAll(baseDir, cacheDirFilePerm)
+	if err != nil {
```
Comment (varas) [inline cache/global.go]: Extra var `err`

File: cache/global.go [hunk 5]
```diff
@@ -1,37 +1,82 @@
...
 var GlobalLog = log.NewStdErr(DefaultDebug)
 
-var instance, err = NewCache(GlobalLog)
-
 func SetupLogging(verbose bool) {
 	GlobalLog.SetDebug(verbose)
 }
 
+const integrationsDir = "nr-integrations"
+const cacheDirFilePerm = 0755
```
Comment (varas) [inline cache/global.go]: i'd group consts before vars like:
```
const (
    constA = 1
    constB ...
)
```

File: cache/global.go [hunk 6]
```diff
@@ -1,37 +1,82 @@
...
 // TODO delete when removing these globals
 var GlobalLog = log.NewStdErr(DefaultDebug)
 
-var instance, err = NewCache(GlobalLog)
-
 func SetupLogging(verbose bool) {
 	GlobalLog.SetDebug(verbose)
 }
 
+const integrationsDir = "nr-integrations"
```
Comment (varas) [inline cache/global.go]: why this new behaviour?

File: cache/global.go [hunk 7]
```diff
@@ -1,37 +1,82 @@
...
+	cachePath := os.Getenv(pathEnvVar)
+	if cachePath == "" {
+		_, fname := filepath.Split(os.Args[0])
+		cachePath = filepath.Join(os.TempDir(), fmt.Sprintf("%s-global.json", fname))
+	}
+	return cachePath
+}
+
+var instance Cache
+var err error
```
Comment (varas) [inline cache/global.go]: grouped on top please

File: sdk/v2/builder.go [hunk 8]
```diff
@@ -26,10 +27,17 @@ type IntegrationBuilder interface {
 	// Writer sets the output stream where the integration resulting payload will be written to.
 	// By default, the standard output (os.Stdout).
 	Writer(io.Writer) IntegrationBuilder
+	// Cache sets the Cache implementation that will be used to persist data between executions of the same integration.
+	// By default, it will be a Disk-backed cache named stored in the file returned by the
+	// cache.DefaultPath(integrationName) function.
+	Cache(cache.Cache) IntegrationBuilder
+	// NoCache disables the cache for this integration.
+	NoCache() IntegrationBuilder
```
Comment (varas) [inline sdk/v2/builder.go]: Not sure about this.

Integrations do need a persistent cache, I guess we should create a one per integration and hide it to the user.

Why a user would like to set it's own cache? or remove the current one?

Comment (mariomac) [reply sdk/v2/builder.go]: I think that integrations without deltas don't need a cache.

Also, they could prefer using other caches, like redis, etcd, etc....

File: cache/cache_test.go (hunk 3 above)
Comment (mariomac) [reply cache/cache_test.go]: To be consistent with the previous work from the team, the reason is that this way you are "forced" to test only against public interfaces.

File: cache/global.go (hunk 6 above)
Comment (mariomac) [reply cache/global.go]: Just to have all the caches in a single directory, in case more than one integration coexist.

**Comment by mariomac** [comment]
Good point, Juan. I'll implement it into a new patch.

**Pull request merged**